<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>The circos.heatmap() function - Sample code rewritten by circlizePlus to generate graphics in "Circular Visualization in R"</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "The circos.heatmap() function";
        var mkdocs_page_input_path = "06-circos-heatmap.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Sample code rewritten by circlizePlus to generate graphics in "Circular Visualization in R"
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Prerequisite</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../01-introduction/">Introduction</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../02-circlize-layout/">Circular layout</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../03-graphics/">Graphics</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../04-legends/">Legends</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../05-implement-high-level-plots/">Implement high-level circular plots</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">The circos.heatmap() function</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#input-data">Input data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#circular-layout">Circular layout</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#dendrograms-and-row-names">Dendrograms and row names</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#clustering">Clustering</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#callback-on-dendrograms">Callback on dendrograms</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#multiple-heatmap-tracks">Multiple heatmap tracks</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#with-other-tracks">With other tracks</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#add-annotations">Add annotations</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#a-complex-example-of-circular-heatmaps">A complex example of circular heatmaps</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../07-advanced-usage/">Advanced layout</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../09-initialize-genomic-data/">Initialize with genomic data</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../10-create-plotting-regions/">Create plotting regions for genomic data</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../11-modes-of-input/">Modes for circos.genomicTrack()</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../12-high-level-genomic-functions/">High-level genomic functions</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../13-nested-zooming/">Nested zooming</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Sample code rewritten by circlizePlus to generate graphics in "Circular Visualization in R"</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">The circos.heatmap() function</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="the-circosheatmap-function">The <code>circos.heatmap()</code> function</h1>
<pre><code class="language-r">set.seed(123)
mat1 = rbind(cbind(matrix(rnorm(50*5, mean = 1), nr = 50), 
                   matrix(rnorm(50*5, mean = -1), nr = 50)),
             cbind(matrix(rnorm(50*5, mean = -1), nr = 50), 
                   matrix(rnorm(50*5, mean = 1), nr = 50))
            )
rownames(mat1) = paste0(&quot;R&quot;, 1:100)
colnames(mat1) = paste0(&quot;C&quot;, 1:10)
mat1 = mat1[sample(100, 100), ] # randomly permute rows
split = sample(letters[1:5], 100, replace = TRUE)
split = factor(split, levels = letters[1:5])
</code></pre>
<pre><code class="language-r">library(ComplexHeatmap)
Heatmap(mat1, row_split = split)
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-2-1.png" alt="A normal heatmap." width="384" />
<p class="caption">A normal heatmap.</p>
</div>

<h2 id="input-data">Input data</h2>
<pre><code class="language-r">library(circlize) # &gt;= 0.4.10
col_fun1 = colorRamp2(c(-2, 0, 2), c(&quot;blue&quot;, &quot;white&quot;, &quot;red&quot;))
ccHeatmap(mat = mat1, split = split, col = col_fun1)
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-3-1.png" alt="A circular heatmap which has been split." width="672" />
<p class="caption">A circular heatmap which has been split.</p>
</div>

<pre><code class="language-r">circos.clear()
</code></pre>
<pre><code class="language-r">ccHeatmap(mat = mat1, col = col_fun1)
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-4-1.png" alt="A circular heatmap which no split." width="672" />
<p class="caption">A circular heatmap which no split.</p>
</div>

<pre><code class="language-r">circos.clear()
</code></pre>
<h2 id="circular-layout">Circular layout</h2>
<pre><code class="language-r">par1 = ccPar(start.degree = 90, gap.degree = 10)
cc = ccHeatmap(mat = mat1, split = split, col = col_fun1, track.height = 0.4, bg.border = &quot;green&quot;, bg.lwd = 2, bg.lty = 2, show.sector.labels = TRUE)
cc + par1
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-5-1.png" alt="Circular heatmap. Control the layout." width="672" />
<p class="caption">Circular heatmap. Control the layout.</p>
</div>

<pre><code class="language-r">circos.clear()
</code></pre>
<pre><code class="language-r"># note since circos.clear() was called in the previous plot,
# now the layout starts from theta = 0 (the first sector is 'e')
ccHeatmap(mat = mat1, split = factor(split, levels = c(&quot;e&quot;, &quot;d&quot;, &quot;c&quot;, &quot;b&quot;, &quot;a&quot;)), 
    col = col_fun1, show.sector.labels = TRUE)
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-6-1.png" alt="Circular heatmap. Control the order of heatmaps." width="672" />
<p class="caption">Circular heatmap. Control the order of heatmaps.</p>
</div>

<pre><code class="language-r">circos.clear()
</code></pre>
<h2 id="dendrograms-and-row-names">Dendrograms and row names</h2>
<pre><code class="language-r">ccHeatmap(mat = mat1, split = split, col = col_fun1, dend.side = &quot;inside&quot;)
circos.clear()
ccHeatmap(mat = mat1, split = split, col = col_fun1, dend.side = &quot;outside&quot;)
circos.clear()
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-8-1.png" alt="Circular heatmap. Control the dendrograms." width="768" />
<p class="caption">Circular heatmap. Control the dendrograms.</p>
</div>

<pre><code class="language-r">ccHeatmap(mat = mat1, split = split, col = col_fun1, rownames.side = &quot;inside&quot;)
circos.clear()
text(0, 0, 'rownames.side = &quot;inside&quot;')
ccHeatmap(mat = mat1, split = split, col = col_fun1, rownames.side = &quot;outside&quot;)
circos.clear()
text(0, 0, 'rownames.side = &quot;outside&quot;')
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-10-1.png" alt="Circular heatmap. Control the row names." width="768" />
<p class="caption">Circular heatmap. Control the row names.</p>
</div>

<pre><code class="language-r">ccHeatmap(mat = mat1, split = split, col = col_fun1, dend.side = &quot;inside&quot;, 
    rownames.side = &quot;outside&quot;)
circos.clear()
ccHeatmap(mat = mat1, split = split, col = col_fun1, dend.side = &quot;outside&quot;, 
    rownames.side = &quot;inside&quot;)
circos.clear()
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-12-1.png" alt="Circular heatmap. Control both the dendrograms and row names." width="768" />
<p class="caption">Circular heatmap. Control both the dendrograms and row names.</p>
</div>

<pre><code class="language-r">ccHeatmap(mat = mat1, split = split, col = col_fun1, rownames.side = &quot;outside&quot;,
    rownames.col = 1:nrow(mat1) %% 10 + 1,
    rownames.cex = runif(nrow(mat1), min = 0.3, max = 2),
    rownames.font = 1:nrow(mat1) %% 4 + 1)
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-13-1.png" alt="Circular heatmap. Control graphic parameters for row names." width="672" />
<p class="caption">Circular heatmap. Control graphic parameters for row names.</p>
</div>

<pre><code class="language-r">circos.clear()
</code></pre>
<h2 id="clustering">Clustering</h2>
<pre><code class="language-r">ccHeatmap(mat = mat1, split = split, cluster = FALSE, col = col_fun1)
circos.clear()
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-15-1.png" alt="Circular heatmap. Control clusterings." width="768" />
<p class="caption">Circular heatmap. Control clusterings.</p>
</div>

<h2 id="callback-on-dendrograms">Callback on dendrograms</h2>
<pre><code class="language-r">function(dend, m, si) reorder(dend, rowMeans(m))
</code></pre>
<pre><code class="language-r">library(dendsort)
ccHeatmap(mat = mat1, split = split, col = col_fun1, dend.side = &quot;inside&quot;,
    dend.callback = function(dend, m, si) {
        dendsort(dend)
    }
)
circos.clear()
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-19-1.png" alt="Circular Heatmap. Reorder dendrograms." width="768" />
<p class="caption">Circular Heatmap. Reorder dendrograms.</p>
</div>

<pre><code class="language-r">library(dendextend)
dend_col = structure(1:5, names = letters[1:5])
ccHeatmap(mat = mat1, split = split, col = col_fun1, dend.side = &quot;inside&quot;,
    dend.track.height = 0.2,
    dend.callback = function(dend, m, si) {
        # when k = 1, it renders one same color for the whole dendrogram
        color_branches(dend, k = 1, col = dend_col[si])
    }
)
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-20-1.png" alt="Circular heatmap. Render dendrograms that were split." width="672" />
<p class="caption">Circular heatmap. Render dendrograms that were split.</p>
</div>

<pre><code class="language-r">circos.clear()
</code></pre>
<pre><code class="language-r">ccHeatmap(mat = mat1, col = col_fun1, dend.side = &quot;inside&quot;,
    dend.track.height = 0.2,
    dend.callback = function(dend, m, si) {
        color_branches(dend, k = 4, col = 2:5)
    }
)
</code></pre>
<div class="figure" style="text-align: center">
<img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-21-1.png" alt="Circular heatmap. Render dendrograms." width="672" />
<p class="caption">Circular heatmap. Render dendrograms.</p>
</div>

<pre><code class="language-r">circos.clear()
</code></pre>
<h2 id="multiple-heatmap-tracks">Multiple heatmap tracks</h2>
<pre><code class="language-r">mat2 = mat1[sample(100, 100), ] # randomly permute mat1 by rows
col_fun2 = colorRamp2(c(-2, 0, 2), c(&quot;green&quot;, &quot;white&quot;, &quot;red&quot;))

hm1 = ccHeatmap(mat = mat1, split = split, col = col_fun1, dend.side = &quot;outside&quot;)
hm2 = ccHeatmap(mat = mat2, col = col_fun2)
hm1 + hm2
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code class="language-r">circos.clear()
</code></pre>
<pre><code class="language-r">hm1 = ccHeatmap(mat = mat2, split = split, col = col_fun2, dend.side = &quot;outside&quot;)
hm2 = ccHeatmap(mat = mat1, col = col_fun1)
hm1 + hm2
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-23-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code class="language-r">circos.clear()
</code></pre>
<pre><code class="language-r">cc = ccPlot(initMode='heatmap.initialize', mat = mat1, split = split)
hm1 = ccHeatmap(mat = mat2, col = col_fun2, dend.side = &quot;outside&quot;)
hm2 = ccHeatmap(mat = mat1, col = col_fun1)
cc + hm1 + hm2
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-24-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code class="language-r">circos.clear()
</code></pre>
<pre><code class="language-r">cc = ccPlot(initMode='heatmap.initialize', mat = mat1, split = split)
hm1 = ccHeatmap(mat = mat1[, 1:5], col = col_fun1)
hm2 = ccHeatmap(mat = mat1[, 6:10], col = col_fun1)
cc + hm1 + hm2
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-25-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code class="language-r">circos.clear()
</code></pre>
<h2 id="with-other-tracks">With other tracks</h2>
<pre><code>CELL_META$row_dend
## 'dendrogram' with 2 branches and 14 members total, at height 10.51736 

CELL_META$row_order
## [1]  2  6  4 12  8  1  5 10  7  9 13 11  3 14

CELL_META$subset
## [1]  8  9 14 18 20 37 55 62 66 72 78 85 93 97
</code></pre>
<pre><code class="language-r">hm1 = ccHeatmap(mat = mat1, split = split, col = col_fun1)
row_mean = rowMeans(mat1[, 1:5])
t1 = ccTrack(ylim = range(row_mean), panel.fun = function(x, y) {
    y = row_mean[CELL_META$subset]
    y = y[CELL_META$row_order]
    circos.lines(CELL_META$cell.xlim, c(0, 0), lty = 2, col = &quot;grey&quot;)
    circos.points(seq_along(y) - 0.5, y, col = ifelse(y &gt; 0, &quot;red&quot;, &quot;blue&quot;))
}, cell.padding = c(0.02, 0, 0.02, 0))
hm1 + t1
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code class="language-r">circos.clear()
</code></pre>
<pre><code class="language-r">cc = ccPlot(initMode='heatmap.initialize', mat = mat1, split = split)
# This is the same as the previous example
t1 = ccTrack(ylim = range(row_mean), panel.fun = function(x, y) {
    y = row_mean[CELL_META$subset]
    y = y[CELL_META$row_order]
    circos.lines(CELL_META$cell.xlim, c(0, 0), lty = 2, col = &quot;grey&quot;)
    circos.points(seq_along(y) - 0.5, y, col = ifelse(y &gt; 0, &quot;red&quot;, &quot;blue&quot;))
}, cell.padding = c(0.02, 0, 0.02, 0))
hm1 = ccHeatmap(mat1, col = col_fun1) # no need to specify 'split' here
cc + t1 + hm1
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-27-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code class="language-r">circos.clear()
</code></pre>
<pre><code class="language-r">hm1 = ccHeatmap(mat = mat1, split = split, col = col_fun1)
t1 = ccTrack(ylim = range(mat1), panel.fun = function(x, y) {
    m = mat1[CELL_META$subset, 1:5, drop = FALSE]
    m = m[CELL_META$row_order, , drop = FALSE]
    n = nrow(m)
    # circos.boxplot is applied on matrix columns, so here we transpose it.
    circos.boxplot(t(m), pos = 1:n - 0.5, pch = 16, cex = 0.3)
    circos.lines(CELL_META$cell.xlim, c(0, 0), lty = 2, col = &quot;grey&quot;)
}, cell.padding = c(0.02, 0, 0.02, 0))
hm1 + t1
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-28-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code class="language-r">circos.clear()
</code></pre>
<h2 id="add-annotations">Add annotations</h2>
<pre><code class="language-r">hm1 = ccHeatmap(mat = mat1, split = split, col = col_fun1)
hm1
circos.track(track.index = get.current.track.index(), panel.fun = function(x, y) {
    circos.text(CELL_META$xcenter, CELL_META$cell.ylim[2] + convert_y(2, &quot;mm&quot;), 
        paste0(&quot;this is group &quot;, CELL_META$sector.index),
        facing = &quot;bending.inside&quot;, cex = 0.8,
        adj = c(0.5, 0), niceFacing = TRUE)
}, bg.border = NA)
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code class="language-r">circos.clear()
</code></pre>
<pre><code class="language-r">par1 = ccPar(gap.after = c(2, 2, 2, 2, 10))
hm1 = ccHeatmap(mat = mat1, split = split, col = col_fun1, track.height = 0.4)
hm1 + par1
circos.track(track.index = get.current.track.index(), panel.fun = function(x, y) {
    if(CELL_META$sector.numeric.index == 5) { # the last sector
        cn = colnames(mat1)
        n = length(cn)
        circos.text(rep(CELL_META$cell.xlim[2], n) + convert_x(1, &quot;mm&quot;), 
            1:n - 0.5, cn, 
            cex = 0.5, adj = c(0, 0.5), facing = &quot;inside&quot;)
    }
}, bg.border = NA)
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-30-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code class="language-r">circos.clear()
</code></pre>
<pre><code class="language-r">par1 = ccPar(gap.after = c(2, 2, 2, 2, 10))
hm1 = ccHeatmap(mat = mat1, split = split, col = col_fun1, track.height = 0.4)
hm1 + par1
circos.track(track.index = get.current.track.index(), panel.fun = function(x, y) {
    if(CELL_META$sector.numeric.index == 5) { # the last sector
        circos.rect(CELL_META$cell.xlim[2] + convert_x(1, &quot;mm&quot;), 0,
                    CELL_META$cell.xlim[2] + convert_x(5, &quot;mm&quot;), 5,
                    col = &quot;orange&quot;, border = NA)
        circos.text(CELL_META$cell.xlim[2] + convert_x(3, &quot;mm&quot;), 2.5,
                    &quot;group 1&quot;, cex = 0.5, facing = &quot;clockwise&quot;)

        circos.rect(CELL_META$cell.xlim[2] + convert_x(1, &quot;mm&quot;), 5,
                    CELL_META$cell.xlim[2] + convert_x(5, &quot;mm&quot;), 10,
                    col = &quot;pink&quot;, border = NA)
        circos.text(CELL_META$cell.xlim[2] + convert_x(3, &quot;mm&quot;), 7.5,
                    &quot;group 2&quot;, cex = 0.5, facing = &quot;clockwise&quot;)
    }
}, bg.border = NA)
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-31-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code class="language-r">circos.clear()
</code></pre>
<pre><code class="language-r">hm1 = ccHeatmap(mat = mat1, split = split, col = col_fun1)
hm1
circos.clear()

library(ComplexHeatmap)
lgd = Legend(title = &quot;mat1&quot;, col_fun = col_fun1)
grid.draw(lgd)
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-32-1.png" width="672" style="display: block; margin: auto;" /></p>
<h2 id="a-complex-example-of-circular-heatmaps">A complex example of circular heatmaps</h2>
<pre><code class="language-r">source(&quot;https://gist.githubusercontent.com/jokergoo/0ea5639ee25a7edae3871ed8252924a1/raw/57ca9426c2ed0cebcffd79db27a024033e5b8d52/random_matrices.R&quot;)
</code></pre>
<pre><code class="language-r">set.seed(123)
km = kmeans(mat_meth, centers = 5)$cluster
</code></pre>
<pre><code class="language-r">col_meth = colorRamp2(c(0, 0.5, 1), c(&quot;blue&quot;, &quot;white&quot;, &quot;red&quot;))
hm1 = ccHeatmap(mat = mat_meth, split = km, col = col_meth, track.height = 0.12)

col_direction = c(&quot;hyper&quot; = &quot;red&quot;, &quot;hypo&quot; = &quot;blue&quot;)
hm2 = ccHeatmap(mat = direction, col = col_direction, track.height = 0.01)

col_expr = colorRamp2(c(-2, 0, 2), c(&quot;green&quot;, &quot;white&quot;, &quot;red&quot;))
hm3 = ccHeatmap(mat = mat_expr, col = col_expr, track.height = 0.12)

col_pvalue = colorRamp2(c(0, 2, 4), c(&quot;white&quot;, &quot;white&quot;, &quot;red&quot;))
hm4 = ccHeatmap(mat = cor_pvalue, col = col_pvalue, track.height = 0.01)

library(RColorBrewer)
col_gene_type = structure(brewer.pal(length(unique(gene_type)), &quot;Set3&quot;), names = unique(gene_type))
hm5 = ccHeatmap(mat = gene_type, col = col_gene_type, track.height = 0.01)

col_anno_gene = structure(brewer.pal(length(unique(anno_gene)), &quot;Set1&quot;), names = unique(anno_gene))
hm6 = ccHeatmap(mat = anno_gene, col = col_anno_gene, track.height = 0.01) 

col_dist = colorRamp2(c(0, 10000), c(&quot;black&quot;, &quot;white&quot;))
hm7 = ccHeatmap(mat = dist, col = col_dist, track.height = 0.01)

col_enhancer = colorRamp2(c(0, 1), c(&quot;white&quot;, &quot;orange&quot;))
hm8 = ccHeatmap(mat = anno_enhancer, col = col_enhancer, track.height = 0.03)
cc = hm1 + hm2 + hm3 + hm4 + hm5 + hm6 + hm7 + hm8
cc
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-35-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre><code class="language-r">circos.clear()
</code></pre>
<pre><code class="language-r">df_link = data.frame(
    from_index = sample(nrow(mat_meth), 20),
    to_index = sample(nrow(mat_meth), 20)
)
</code></pre>
<pre><code class="language-r">for(i in seq_len(nrow(df_link))) {
    # Let's call the DMR with index df_link$from_index[i] as DMR1,
    # and the other one with index df_link$to_index[i] as DMR2.

    # The sector where DMR1 is in.
    group1 = km[ df_link$from_index[i] ]
    # The sector where DMR2 is in.
    group2 = km[ df_link$to_index[i] ]

    # The subset of DMRs (row indices from mat_meth) in sector `group1`.
    subset1 = get.cell.meta.data(&quot;subset&quot;, sector.index = group1)
    # The row ordering in sector `group1`.
    row_order1 = get.cell.meta.data(&quot;row_order&quot;, sector.index = group1)
    # This is the position of DMR1 in the `group1` heatmap.
    x1 = which(subset1[row_order1] == df_link$from_index[i])

    # The subset of DMRs (row indices from mat_meth) in sector `group2`.
    subset2 = get.cell.meta.data(&quot;subset&quot;, sector.index = group2)
    # The row ordering in sector `group2`.
    row_order2 = get.cell.meta.data(&quot;row_order&quot;, sector.index = group2)
    # This is the position of DMR2 in the `group2` heatmap.
    x2 = which(subset2[row_order2] == df_link$to_index[i])

    # We take the middle point and draw a link between DMR1 and DMR2
    cc = cc + ccLink(group1, x1 - 0.5, group2, x2 - 0.5, col = rand_color(1))
}
cc
</code></pre>
<pre><code class="language-r">for(i in seq_len(nrow(df_link))) {
    cc = cc + ccHeatmapLink(df_link$from_index[i],
                        df_link$to_index[i],
                        col = rand_color(1))
}
cc
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-39-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre><code class="language-r">circlize_plot = function() {
    hm1 = ccHeatmap(mat = mat_meth, split = km, col = col_meth, track.height = 0.12)
    hm2 = ccHeatmap(mat = direction, col = col_direction, track.height = 0.01)
    hm3 = ccHeatmap(mat = mat_expr, col = col_expr, track.height = 0.12)
    hm4 = ccHeatmap(mat = cor_pvalue, col = col_pvalue, track.height = 0.01)
    hm5 = ccHeatmap(mat = gene_type, col = col_gene_type, track.height = 0.01)
    hm6 = ccHeatmap(mat = anno_gene, col = col_anno_gene, track.height = 0.01) 
    hm7 = ccHeatmap(mat = dist, col = col_dist, track.height = 0.01)
    hm8 = ccHeatmap(mat = anno_enhancer, col = col_enhancer, track.height = 0.03)
    cc = hm1 + hm2 + hm3 + hm4 + hm5 + hm6 + hm7 + hm8
    for(i in seq_len(nrow(df_link))) {
        cc = cc + ccHeatmapLink(df_link$from_index[i],
                        df_link$to_index[i],
                        col = rand_color(1))
    }
    show(cc)    
    circos.clear()

}
</code></pre>
<pre><code class="language-r">lgd_meth = Legend(title = &quot;Methylation&quot;, col_fun = col_meth)
lgd_direction = Legend(title = &quot;Direction&quot;, at = names(col_direction), 
    legend_gp = gpar(fill = col_direction))
lgd_expr = Legend(title = &quot;Expression&quot;, col_fun = col_expr)
lgd_pvalue = Legend(title = &quot;P-value&quot;, col_fun = col_pvalue, at = c(0, 2, 4), 
    labels = c(1, 0.01, 0.0001))
lgd_gene_type = Legend(title = &quot;Gene type&quot;, at = names(col_gene_type), 
    legend_gp = gpar(fill = col_gene_type))
lgd_anno_gene = Legend(title = &quot;Gene anno&quot;, at = names(col_anno_gene), 
    legend_gp = gpar(fill = col_anno_gene))
lgd_dist = Legend(title = &quot;Dist to TSS&quot;, col_fun = col_dist, 
    at = c(0, 5000, 10000), labels = c(&quot;0kb&quot;, &quot;5kb&quot;, &quot;10kb&quot;))
lgd_enhancer = Legend(title = &quot;Enhancer overlap&quot;, col_fun = col_enhancer, 
    at = c(0, 0.25, 0.5, 0.75, 1), labels = c(&quot;0%&quot;, &quot;25%&quot;, &quot;50%&quot;, &quot;75%&quot;, &quot;100%&quot;))
</code></pre>
<pre><code class="language-r">library(gridBase)
plot.new()
circle_size = unit(1, &quot;snpc&quot;) # snpc unit gives you a square region

pushViewport(viewport(x = 0, y = 0.5, width = circle_size, height = circle_size,
    just = c(&quot;left&quot;, &quot;center&quot;)))
par(omi = gridOMI(), new = TRUE)
circlize_plot()
upViewport()

h = dev.size()[2]
lgd_list = packLegend(lgd_meth, lgd_direction, lgd_expr, lgd_pvalue, lgd_gene_type, 
    lgd_anno_gene, lgd_dist, lgd_enhancer, max_height = unit(0.9*h, &quot;inch&quot;))
draw(lgd_list, x = circle_size, just = &quot;left&quot;)
</code></pre>
<p><img src="../06-circos-heatmap_files/figure-html/unnamed-chunk-42-1.png" width="100%" style="display: block; margin: auto;" /></p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../05-implement-high-level-plots/" class="btn btn-neutral float-left" title="Implement high-level circular plots"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../07-advanced-usage/" class="btn btn-neutral float-right" title="Advanced layout">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../05-implement-high-level-plots/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../07-advanced-usage/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
